test/test_autograd.py

```diff
@@ -9019,19 +9052,38 @@ def train_fn_grad(x):
         # be accumulate to the same place and should be the same
         self._run_py_multithread_fn(train_fn_grad, (x,))
 
-    def test_multithread_saved_tensors_hooks(self):
-        def pack(x):
-            warnings.warn("pack")
-            return x
+    def test_multi_grad_hooks(self):
+        # Multihooks should behave independently per execution of backward
+        # Test that the hook fired the number of times we ran backward
+        # even if those executions occur concurrently on different threads
+        t1 = torch.rand(2, requires_grad=True)
+        t2 = torch.rand(2, requires_grad=True)
+        t3 = torch.rand(2, requires_grad=True)
+        t4 = torch.rand(2, requires_grad=True)
+
+        res = None
+        count = [0]
+
+        def hook(grads):
+            nonlocal res
+            count[0] += 1
+            grad_is_none = [g is not None for g in grads]
+            if res is None:
+                res = grad_is_none
+            else:
+                self.assertEqual(res, grad_is_none)
 
-        def registers_hooks_for_each_thread():
-            with torch.autograd.graph.saved_tensors_hooks(pack, lambda x: x):
-                x = torch.ones(5, 5, requires_grad=True)
-                with warnings.catch_warnings(record=True) as w:
-                    y = x * x
-                    # should raise two warnings from x being saved twice
-                    self.assertEqual(len(w), 2)
-            y.sum().backward()
```

Comment:
I removed this because it didn't actually run any code previously, and I tried modifying it to try to get it work, but the number of warnings produced seems to be non-deterministic.

```python
def test_multithread_saved_tensors_hooks(self):
      def pack(x):
          warnings.warn("pack")
          return x

      num_warnings = [0]
      def registers_hooks_for_each_thread():
          warnings.simplefilter("always")
          with torch.autograd.graph.saved_tensors_hooks(pack, lambda x: x):
              x = torch.ones(5, 5, requires_grad=True)
              with warnings.catch_warnings(record=True) as w:
                  y = x * x
              num_warnings[0] += len(w)

      with warnings.catch_warnings(record=True) as w:
          self._run_py_multithread_fn(registers_hooks_for_each_thread, num_threads=10)

      self.assertEqual(num_warnings[0] + len(w), 20)
```